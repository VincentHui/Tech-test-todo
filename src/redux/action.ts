import { TodoActionsTypes, Todo, CREATE_TODO_REQUEST, UPDATE_TODO_REQUEST, DELETE_TODO_REQUEST } from './types';
let nextTodoId = 0;

export const getIdIncrementer = (): number => nextTodoId;

export const createTodo = (description: string): TodoActionsTypes => {
    nextTodoId++;
    return {
        type: CREATE_TODO_REQUEST,
        payload: { todo: { description: description, checked: false, id: nextTodoId } },
    };
};
export const updateTodo = (todo: Todo): TodoActionsTypes => {
    return {
        type: UPDATE_TODO_REQUEST,
        payload: { todo },
    };
};
export const removeTodo = (todoId: number): TodoActionsTypes => {
    return {
        type: DELETE_TODO_REQUEST,
        id: todoId,
    };
};
