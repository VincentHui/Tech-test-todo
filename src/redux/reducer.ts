/* reducer.ts */

import { TodoState, TodoActionsTypes, CREATE_TODO_REQUEST, UPDATE_TODO_REQUEST, DELETE_TODO_REQUEST } from './types';

const initialState: TodoState = {
    data: [
        { id: -1, description: 'make tests', checked: false },
        { id: -2, description: 'make store logic', checked: false },
        { id: -3, description: 'make ui', checked: true },
    ],
};

export default function todoReducer(state = initialState, action: TodoActionsTypes): TodoState {
    switch (action.type) {
        case CREATE_TODO_REQUEST:
            return {
                data: [...state.data, action.payload.todo],
            };
        case UPDATE_TODO_REQUEST:
            return {
                data: state.data.map((todo) => (todo.id === action.payload.todo.id ? action.payload.todo : todo)),
            };
        case DELETE_TODO_REQUEST:
            return {
                data: state.data.filter((todo) => todo.id !== action.id),
            };
        default:
            return state;
    }
}
