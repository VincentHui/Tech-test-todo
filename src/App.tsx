/* eslint-disable @typescript-eslint/no-empty-function */
import React, { useState } from 'react';
import { useSpring, animated, config } from 'react-spring';
import styled from 'styled-components';
import { useSelector, useDispatch } from 'react-redux';
import { TodoState, Todo } from './redux/types';
import { createTodo, updateTodo, removeTodo } from './redux/action';
const List = styled.ul`
    list-style: none;
`;

const Row = styled.div`
    display: flex;
    flex-direction: row;
    align-items: center;
`;

const ToDoInput = styled.input`
    margin-right: 20px;
`;

const TodoButton = styled.button`
    min-height: 20px;
`;

const TodoHeading = styled.h2<{ checked?: boolean }>`
    width: 300px;
    font-size: larger;
    flex-grow: 2;
    overflow-wrap: break-word;
    word-wrap: break-word;
    hyphens: auto;
    ${(props) => (!props.checked ? '' : 'text-decoration:line-through')};
`;

const TodoSection = styled.section`
    padding: 40px;
    display: flex;
    flex-direction: column;
`;
interface todoRowProp {
    id: number;
    description: string;
    checked: boolean;
}

export const TodoRow: React.FC<todoRowProp> = (props: todoRowProp) => {
    const dispatch = useDispatch();
    const [editValue, setEditValue] = useState(false);
    const [hover, setHover] = useState(false);
    const hoverSpring = useSpring({
        to: { transform: hover && !editValue ? 'scale(1.02)' : 'scale(1)' },
        from: { transform: 'scale(1)' },
    });
    const onEdit = (value: string) => {
        setEditValue(false);
        value.length === 0 ? dispatch(removeTodo(props.id)) : dispatch(updateTodo({ ...props, description: value }));
    };
    return (
        <animated.div style={{ ...hoverSpring, display: 'flex', minHeight: 70 }}>
            <Row onMouseOver={() => setHover(true)} onMouseOut={() => setHover(false)}>
                <InputForm
                    clearinput={false}
                    hidden={!editValue}
                    buttonName="update"
                    label="button to update item"
                    placeHolder={props.description}
                    submitCallBack={onEdit}
                    blurCallback={onEdit}
                />

                <Row style={{ display: editValue ? 'none' : 'flex' }}>
                    <ToDoInput
                        type="checkbox"
                        defaultChecked={props.checked}
                        onChange={(e) => dispatch(updateTodo({ ...props, checked: e.target.checked }))}
                    />
                    <TodoHeading checked={props.checked} onClick={() => setEditValue(true)}>
                        {props.description}
                    </TodoHeading>
                    <TodoButton aria-label="Remove a todo item" onClick={() => dispatch(removeTodo(props.id))}>
                        remove
                    </TodoButton>
                </Row>
            </Row>
        </animated.div>
    );
};

export const AddRow: React.FC = () => {
    const dispatch = useDispatch();
    const placeHolder = 'E.g. do a tech test';
    return (
        <Row>
            <InputForm
                buttonName="Add"
                label="button to add todo item"
                placeHolder={placeHolder}
                blurCallback={(e) => {}}
                submitCallBack={(value) => dispatch(createTodo(value))}
            />
        </Row>
    );
};

export interface inputFormProp {
    placeHolder: string;
    submitCallBack: (value: string) => void;
    blurCallback: (value: string) => void;
    buttonName: string;
    label: string;
    hidden?: boolean;
    clearinput?: boolean;
}
export const InputForm: React.FC<inputFormProp> = (props: inputFormProp) => {
    const [inputValue, setInput] = useState(props.clearinput ? '' : props.placeHolder);
    return (
        <form style={{ display: props.hidden ? 'none' : 'block' }}>
            <ToDoInput
                autoFocus
                type="text"
                aria-label={props.label}
                placeholder={props.placeHolder}
                value={inputValue}
                onChange={(e) => setInput(e.target.value)}
                onBlur={(e) => {
                    props.blurCallback(e.target.value);
                }}
            />
            <TodoButton
                type="submit"
                onClick={(e) => {
                    e.preventDefault();
                    props.submitCallBack(inputValue);
                    props.clearinput && setInput('');
                }}
            >
                {props.buttonName}
            </TodoButton>
        </form>
    );
};
InputForm.defaultProps = {
    hidden: false,
    clearinput: true,
};
export const App: React.FC = () => {
    const springProps = useSpring({
        config: config.molasses,
        to: { opacity: 1, transform: 'translateY(0px)' },
        from: { opacity: 0, transform: 'translateY(-50px)' },
    });
    const todoList = useSelector<TodoState, Todo[]>((state) => state.data);
    return (
        <animated.div style={springProps} className="App">
            <TodoSection aria-labelledby="todos-label">
                <h1 id="todos-label">Todo List</h1>
                <List>
                    {todoList.map((todo) => {
                        return (
                            <li key={todo.id}>
                                <TodoRow id={todo.id} description={todo.description} checked={todo.checked} />
                            </li>
                        );
                    })}
                </List>
                <AddRow />
            </TodoSection>
        </animated.div>
    );
};
