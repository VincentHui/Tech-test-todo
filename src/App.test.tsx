/* eslint-disable @typescript-eslint/no-empty-function */
import React from 'react';
import { cleanup, fireEvent, render, screen } from '@testing-library/react';
import { InputForm, TodoRow, AddRow, App } from './App';
import { createTodo, updateTodo, removeTodo, getIdIncrementer } from './redux/action';
import { CREATE_TODO_REQUEST, DELETE_TODO_REQUEST, UPDATE_TODO_REQUEST } from './redux/types';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import todoReducer from './redux/reducer';
import renderer from 'react-test-renderer';

describe('snapshot tests', () => {
    it('InputForm', () => {
        const link = renderer
            .create(
                <InputForm
                    blurCallback={(e) => {}}
                    submitCallBack={(e) => {}}
                    buttonName="test button"
                    label="test label"
                    placeHolder={'test placeholder'}
                />,
            )
            .toJSON();
        expect(link).toMatchSnapshot();
    });
    it('todoRow', () => {
        const store = createStore(todoReducer);
        const todoRow = renderer
            .create(
                <Provider store={store}>
                    <TodoRow id={1} description="test description" checked={false} />
                </Provider>,
            )
            .toJSON();
        expect(todoRow).toMatchSnapshot();
    });
    it('AddRow', () => {
        const store = createStore(todoReducer);
        const addRow = renderer
            .create(
                <Provider store={store}>
                    <AddRow />
                </Provider>,
            )
            .toJSON();
        expect(addRow).toMatchSnapshot();
    });
});

describe('input form test', () => {
    it('InputForm', () => {
        const result = render(
            <InputForm
                clearinput={false}
                blurCallback={(e) => {}}
                submitCallBack={(e) => {}}
                buttonName="test button"
                label="test label"
                placeHolder={'test placeholder'}
            />,
        );

        const input = result.getByLabelText('test label') as HTMLInputElement;
        fireEvent.change(input, { target: { value: 'test input' } });
        fireEvent.blur(input);
        expect(input.value).toBe('test input');
        const submit = result.getByText('test button') as HTMLButtonElement;
        fireEvent.change(input, { target: { value: 'change for submission' } });
        fireEvent.click(submit);
        expect(input.value).toBe('change for submission');
    });
});

describe('actions', () => {
    it('add todo action', () => {
        const description = 'text description';
        const id = getIdIncrementer() + 1;
        const expectedAction = {
            type: CREATE_TODO_REQUEST,
            payload: { todo: { description: description, checked: false, id: id } },
        };
        expect(createTodo(description)).toEqual(expectedAction);
    });
    it('remove todo action', () => {
        const id = 2;
        const expectedAction = {
            type: DELETE_TODO_REQUEST,
            id: 2,
        };
        expect(removeTodo(id)).toEqual(expectedAction);
    });
    it('update todo action', () => {
        const todoUpdate = { id: 1, checked: false, description: 'update' };
        const expectedAction = {
            type: UPDATE_TODO_REQUEST,
            payload: { todo: todoUpdate },
        };
        expect(updateTodo(todoUpdate)).toEqual(expectedAction);
    });
});

describe('connected store', () => {
    it('todo logic', () => {
        const testStore = createStore(todoReducer);
        const testDescription = 'test description';
        testStore.dispatch(createTodo('test description'));
        render(
            <Provider store={testStore}>
                <App />
            </Provider>,
        );
        expect(screen.getByText(testDescription)).toBeInTheDocument();
        const createdToDo = testStore.getState().data.filter((todo) => todo.description === testDescription)[0];
        testStore.dispatch(updateTodo({ id: createdToDo.id, checked: true, description: 'updated' }));
        expect(screen.getByText('updated')).toBeInTheDocument();
        testStore.dispatch(removeTodo(createdToDo.id));
        expect(screen.queryByText('updated') === null);
    });
});
