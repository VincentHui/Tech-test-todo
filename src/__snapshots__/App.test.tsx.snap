// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot tests AddRow 1`] = `
<div
  className="sc-AxirZ dAVawm"
>
  <form
    style={
      Object {
        "display": "block",
      }
    }
  >
    <input
      aria-label="button to add todo item"
      autoFocus={true}
      className="sc-AxiKw gVLkSS"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="E.g. do a tech test"
      type="text"
      value=""
    />
    <button
      className="sc-AxhCb jnEOsd"
      onClick={[Function]}
      type="submit"
    >
      Add
    </button>
  </form>
</div>
`;

exports[`snapshot tests InputForm 1`] = `
<form
  style={
    Object {
      "display": "block",
    }
  }
>
  <input
    aria-label="test label"
    autoFocus={true}
    className="sc-AxiKw gVLkSS"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="test placeholder"
    type="text"
    value=""
  />
  <button
    className="sc-AxhCb jnEOsd"
    onClick={[Function]}
    type="submit"
  >
    test button
  </button>
</form>
`;

exports[`snapshot tests todoRow 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "minHeight": 70,
      "transform": " scale(1)",
    }
  }
>
  <div
    className="sc-AxirZ dAVawm"
    onMouseOut={[Function]}
    onMouseOver={[Function]}
  >
    <form
      style={
        Object {
          "display": "none",
        }
      }
    >
      <input
        aria-label="button to update item"
        autoFocus={true}
        className="sc-AxiKw gVLkSS"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="test description"
        type="text"
        value="test description"
      />
      <button
        className="sc-AxhCb jnEOsd"
        onClick={[Function]}
        type="submit"
      >
        update
      </button>
    </form>
    <div
      className="sc-AxirZ dAVawm"
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <input
        className="sc-AxiKw gVLkSS"
        defaultChecked={false}
        onChange={[Function]}
        type="checkbox"
      />
      <h2
        checked={false}
        className="sc-AxhUy iXfeIL"
        onClick={[Function]}
      >
        test description
      </h2>
      <button
        aria-label="Remove a todo item"
        className="sc-AxhCb jnEOsd"
        onClick={[Function]}
      >
        remove
      </button>
    </div>
  </div>
</div>
`;
